src/Core/Services/Implementations/InMemoryApplicationCacheService.cs

```diff
@@ -51,52 +51,89 @@ public virtual async Task<IDictionary<Guid, ProviderAbility>> GetProviderAbiliti
     public virtual async Task UpsertProviderAbilityAsync(Provider provider)
     {
         await InitProviderAbilitiesAsync();
-        var newAbility = new ProviderAbility(provider);
-
-        _providerAbilities[provider.Id] = newAbility;
+        _providerAbilities.AddOrUpdate(
+            provider.Id,
+            _ => new ProviderAbility(provider),
+            (_, _) => new ProviderAbility(provider));
     }
 
     public virtual async Task UpsertOrganizationAbilityAsync(Organization organization)
     {
         await InitOrganizationAbilitiesAsync();
-        var newAbility = new OrganizationAbility(organization);
 
-        _orgAbilities[organization.Id] = newAbility;
+        _orgAbilities.AddOrUpdate(
+            organization.Id,
+            _ => new OrganizationAbility(organization),
+            (_, _) => new OrganizationAbility(organization));
     }
 
     public virtual Task DeleteOrganizationAbilityAsync(Guid organizationId)
     {
-        _orgAbilities?.Remove(organizationId);
-
-        return Task.FromResult(0);
+        _orgAbilities.TryRemove(organizationId, out _);
+        return Task.CompletedTask;
     }
 
     public virtual Task DeleteProviderAbilityAsync(Guid providerId)
     {
-        _providerAbilities?.Remove(providerId);
-
-        return Task.FromResult(0);
+        _providerAbilities.TryRemove(providerId, out _);
+        return Task.CompletedTask;
     }
 
-    private async Task InitOrganizationAbilitiesAsync()
+    private async Task InitOrganizationAbilitiesAsync() =>
+        await InitAbilitiesAsync(
+            () => _orgAbilities,
+            dict => _orgAbilities = dict,
+            () => _lastOrgAbilityRefresh,
+            dt => _lastOrgAbilityRefresh = dt,
+            _orgInitLock,
+            async () => await _organizationRepository.GetManyAbilitiesAsync(),
+            _refreshInterval,
+            ability => ability.Id);
+
+    private async Task InitProviderAbilitiesAsync() =>
```

Comment:
Another approach is to eagerly initialize the two ability ConcurrentDictionary instances during class initialization, since it's guaranteed to run only once. This shifts the initialization from runtime to startup, which avoids the one-time load delay on the first user request (Justin's point in the Slack thread) and also removes the complexity of creating our own locking mechanism.

Reply:
Please disregard this. We’ll still need the double lock for the refresh.